packages/@uppy/dashboard/src/index.js

```diff
@@ -656,6 +657,52 @@ module.exports = class Dashboard extends Plugin {
     this.superFocusOnEachUpdate()
   }
 
+  startUpload = (ev) => {
+    this.uppy.upload().catch((err) => {
+      // Log error.
+      this.uppy.log(err.stack || err.message || err)
+    })
+  }
+
+  cancelUpload = (fileID) => {
+    this.uppy.removeFile(fileID)
+  }
+
+  saveFileCard = (meta, fileID) => {
+    this.uppy.setFileMeta(fileID, meta)
+    this.toggleFileCard()
+  }
+
+  _attachRenderFunctionToTarget = (target) => {
+    const plugin = this.uppy.getPlugin(target.id)
+    return {
+      ...target,
+      icon: plugin.icon || this.opts.defaultPickerIcon,
+      render: plugin.render
+    }
+  }
+
+  _isTargetSupported = (target) => {
+    const plugin = this.uppy.getPlugin(target.id)
+    // If the plugin does not provide a `supported` check, assume the plugin works everywhere.
+    if (typeof plugin.isSupported !== 'function') {
+      return true
+    }
+    return plugin.isSupported()
+  }
+
+  _getAcquirers = memoize((targets) => {
+    return targets
+      .filter(target => target.type === 'acquirer' && this._isTargetSupported(target))
+      .map(this._attachRenderFunctionToTarget)
+  })
+
+  _getProgressIndicators = memoize((targets) => {
+    return targets
+      .filter(target => target.type === 'progressindicator')
+      .map(this._attachRenderFunctionToTarget)
+  })
```

Comment:
Is using `memoize` here basically sort of a hack, because we can’t get access to props, since render is passed the whole `state` in this case?

Reply:
Not really I think, it's just to make sure that this array is not recomputed every render, so that we can use shallowEqual